### ISSUE TITLE
slendr has trouble exporting conda paths

### ISSUE THREAD

**Issue by janxkoci** [bodkan/slendr#154]
When trying to parallelize `slendr` simulations, I run into the same issue using several different methods, which probably boil down to how `slendr` exports its `conda` paths.

In particular, I'm getting the infamous `tskit not found` errors, whenever I try to:

- submit `slendr` scripts as jobs to PBS Pro queue system at MetaCentrum, the Czech national computing grid, or
- use `slendr` in parallel with the [`future` ecosystem](https://www.futureverse.org/).

At the moment, I'm using GNU `parallel` to run simulations simultaneously, but it still doesn't work at MetaCentrum at all (the assigned compute nodes simply don't see the paths, even if I export them explicitly in the script, or if I use an interactive job).

As I mentioned [elsewhere](https://github.com/bodkan/slendr/issues/150#issuecomment-1995571634), `slendr` creates 3 distinct environments, which may be part of the reason why PBS Pro has problems tracking all the paths. But I get the same problem from inside R with active `slendr` environment when I try to use `future` packages, so the problem may be on the `slendr`'s side, when its exporting paths to its tools.

Please, let me know if there is anything I can test or log, to help with resolving the issue. I use both R and `conda` (or `mamba`) pretty much daily, but I'm not very useful when it comes to Python.

**Comment by bodkan** [comment]
OK, I think I know what might be going on. This is indeed similar to other issues so I'll try to describe what's going on in a bit more detail because it might be helpful for other users of reticulate. I could even imagine some poor non-slendr soul searching for the same errors on Google landing here.

But first it might be useful to clarify some terminology:

- The slendr R package doesn't have trouble exporting any paths. The concept of "exporting a path" from unix doesn't apply here, as this isn't something slendr does (conda paths or general Python paths). 
- In fact, this error isn't actually not about slendr at all -- this is an issue of which Python interpreter that's embedded _into_ R when any reticulate code attempts to run any Python code, and applies to any reticulate Python script/function that would be run. The error which happens here is out of slendr's sphere of influence, so to speak.
- In other words, you would have the same issues with any other package which uses Python in R.

I will get to the point (I'm also a heavy user of the _future_ package and am developing a package which uses slendr to runs thousands of parallelized simulations so I recognize your problem) but first I want to establish some background.

**TL;DR:** There's no _"problem on the slendr's side, when its exporting paths to its tools"_ because it's not doing any exporting of any paths. I think you're not running future correctly in this setting. But more on that after the example.

## Example

To explain the above in concrete terms, it's worth following this along, in particularly for a non-expert Python user. Let's forget about slendr for a minute.

Let's say I created the following Python virtual environment. I don't have conda installed on my systems, but the issue isn't conda-related, it's Python-related, so it's more straightforward to explain without bringing conda into picture. As I explained yesterday, slendr only uses conda to fetch the `python.exe` / `python` binary across Windows/Linux/macOS.

First let's verify that a base Python doesn't have matplotlib installed. I chose matplotlib because it's easy to install and to show the issue is general.

```
$ python3
>>> import matplotlib         # matplotlib doesn't work!
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'matplotlib'
```

Now let's create a tiny Python virtual environment which _does_ have matplotlib installed:

```
$ cd /tmp
$ python3 -m venv testenv        # create a virtual environment
$ source testenv/bin/activate    # activate the environment
(testenv) $ python3 -m pip install matplotlib

$ python3
>>> import matplotlib
>>> matplotlib.__version__   # matplotlib works!
'3.8.3'
```

Now deactivate the test Python virtual environment to get a plain shell:

```
(testenv) $ which python3             
/private/tmp/testenv/bin/python3  # python3 interpreter with matplotlib is currently used
(testenv) $ deactivate
$ # now we're no longer in the Python virtual environment
$ which python3
/usr/local/bin/python3  # indeed, after deactivating, the default python3 interpreter is used
```

**Let's now start R and try to import matplotlib (as an example of something "Python related" analogous to the "infamous tskit error"). No conda, no slendr, nothing!** This is just to demonstrate that the issue you're encountering lies deeper than slendr, for the purposes of better understanding this problem.

```
$ R --quiet
> library(reticulate)
> import("matplotlib")
Error in py_module_import(module, convert = convert) : 
  ModuleNotFoundError: No module named 'matplotlib'     # <== the infamous error
Run `reticulate::py_last_error()` for details.
```

You see the "infamous tskit error", as you called it. This happens because the Python interpreter which was embedded into R the moment we executed some Python code (`import("matplotlib")` in R, i.e. `import matplotlib ` in Python) is one which doesn't have msprime installed (typically some "default Python" on the system).

Now, let's close R altogether, and start it again. But this time we instruct R to pick up a Python environment we care about:

```
$ R --quiet
> library(reticulate)
> use_virtualenv("/tmp/testenv")   # this is what `init_env()` does internally -- this and nothing else
> import("matplotlib")
Module(matplotlib)
```

Voilla, beautiful. Stuff works! And the reason it works is that we **first activated a Python virtual environment before doing anything Python-related** (either `venv` or `conda` environment, doesn't matter at all which one). If the Python environment of slendr isn't properly activated first by the R session, there's nothing that can be done by slendr.


## Lessons for slendr

The lesson from above is that **if an R code or an R package requires a Python virtual environment, it must be available and activated before any call to any Python is performed by the script!** If that doesn't happen, R automatically activates whatever Python is available by default, and then calling something like slendr's `msprime()` function simply fails with that "infamous error".


### slendr specific example

I use _future_ and its companion packages all the time, so let's look at an example of how the above Python-vs-reticulate-vs-R embedding issue applies to running stuff in parallel. I will demonstrate running parallelized slendr simulations on a single machine in a sequential, multicore, and multisession settings (I will assume you are [familiar with those](https://cran.r-project.org/web/packages/future/vignettes/future-1-overview.html) as you said you work with those packages yourself).

I am not showing how to run this parallelized across different machines because the same applies -- you just need to make sure that slendr is installed on each remote machine and that `setup_env()` has been created on each of them, of course. Also, I use `future_map_dfr`, but any of the `future_lapply`, etc. work the same, of course.


Let's say I have this setup in which I want to parallelize the computation of nucleotide diversity using slendr. I'll demonstrate this on four scenarios, one of which fails in the way you describe.

```
library(slendr)
init_env(quiet = TRUE) # this internally calls `reticulate::use_*env() as above`

library(furrr) # future iterative wrappers

run_sim <- function(rep) {
  pop <- population("pop", time = 1000, N = 1000)
  model <- compile_model(pop, generation_time = 1, direction = "backward")
  ts <- msprime(model, sequence_length = 10000, recombination_rate = 1e-8)
  pi_df <- ts_diversity(ts, sample_sets = ts_names(ts, split = "pop"), mode = "branch")
  pi_df$rep <- rep
  pi_df
}
```

#### Scenario I. -- sequential, works fine

Nothing to see here. Because `init_env()` has been called in this R session, `run_sim()` can pick it up because it runs in the same process.

```
plan(sequential)
pi_results_I <- future_map_dfr(1:10, run_sim, .options = furrr_options(seed = TRUE))
```


#### Scenario II. -- multicore, works fine

Mode "multicore" runs futures in a shared memory.

This works because futures are evaluated in a shared memory forked from the current R process, which means that the forked processes share the Python interpreter embedded via the call to `init_env()` above into the "parent process".

```
plan(multicore, workers = 10)
pi_results_II <- future_map_dfr(1:10, run_sim, .options = furrr_options(seed = TRUE))
```

#### Scenario III. -- multisession, FAILS!

Executes `run_sim()` in a multisession mode -- this gives the error!

```
import tskit
ModuleNotFoundError: No module named 'tskit'
```

This fails because futures are evaluated in independent processes! Those obviously can't share the embedded Python environment of the "parent R process" because, well, they are independent. As I showed above, if no Python virtual environment is initialized explicitly, whatever Python is lying around by default is used and embedded into the R session.

```
plan(multisession, workers = 10)
pi_results_III <- future_map_dfr(1:10, run_sim, .options = furrr_options(seed = TRUE))
```

#### Scenario IV. -- multisession, FIXED!

This is the same as scenario III. above, but no longer gives the error. Although each future is again evaluated in an independent process like those in scenario III., this works because we initialize the correct embedded Python in each parallelized future independently.

```
plan(multisession, workers = 10)

run_sim_modified <- function(rep) {
  init_env(quiet = TRUE)    # <== note this change!
  pop <- population("pop", time = 1000, N = 1000)
  model <- compile_model(pop, generation_time = 1, direction = "backward")
  ts <- msprime(model, sequence_length = 10000, recombination_rate = 1e-8)
  pi_df <- ts_diversity(ts, sample_sets = ts_names(ts, split = "pop"), mode = "branch")
  pi_df$rep <- rep
  pi_df
}
pi_results_IV <- future_map_dfr(1:10, run_sim_modified, .options = furrr_options(seed = TRUE))
```

## Conclusion

I hope the long (sorry!) write up clears things up a little. I suggest you play around with the examples I provided and try to relate the conclusions from them to your own situation.

In particular, the fact that any independent R process which uses slendr (and, by extension, it's dedicated Python environment) **must activate that Python environment before any Python function is called**. If you don't do this, by that point you try to simulate something it's already too late, because R embedded some default Python which is unlikely to have msprime/tskit/pyslim installed. So, any piece of slendr code that is to be run in parallel must call `init_env()` before anything else, particularly code which is run in a parallelized way across different machines in independent processes.

As an example, I'm working on an ABC package for slendr / SLiM / msprime which involves running potentially millions of parallelized simulations across cores at different machines using futures. Here's [an example of an internal simulation function](https://github.com/bodkan/demografr/blob/c656f88c32c643fa0790343d641ed526097da5da/R/run_iteration.R#L7) (potentially ran across different computers in independent processes) in that project which, In order to avoid the infamous error, makes sure a dedicated slendr Python environment is always activated before anything else.

**Comment by bhaller** [comment]
Wow, what a writeup, @bodkan!  Make sure to put some of this info into your doc!

**Comment by bodkan** [comment]
That's a great point, @bhaller. Thanks. I should start a F.A.Q. vignette with this being item number 1.

This issue luckily doesn't come as often as it used to -- it was way worse back when slendr was trying to activate the built-in Python virtual environment _during_ the `library(slendr)` call. Any time user's Python stuff was misconfigured, it pulled the rug from under slendr without nothing I could do about it.

Splitting the _loading_ of slendr (`library(slendr)`) and initiating a Python environment (`init_env()`) into two individual explicit steps solved most of the obscure low-level issues. I wasn't happy about making this change at first (most of my users don't actually know what Python really is and how it works) but it turned out to be a nice example of "explicit is (sometimes) better than implicit".

**Comment by bhaller** [comment]
The next time I have Python issues, I know who I'm gonna ask ;-)

**Comment by janxkoci** [comment]
Wow, thanks for this detailed reply, @bodkan! I don't recall all the details of my attempts (I'll check the scripts at the HPCs when I have time), but most likely I used `multisession` without any special treatment, which leads to the error. I've found `multisession` to be more reliable than `multicore` on older systems like CentOS7, which are common at HPCs.

**Comment by bodkan** [comment]
Yes, especially when you say that you probably used the "multisession" mode, I'm pretty much certain you were running the simulations without activating the Python environment first. This is the only situation the error can happen because it comes from [slendr evaluating `import tskit`](https://github.com/bodkan/slendr/blob/398bb7af6a6585dcebad6a5b69c07d588659743c/R/zzz.R#L74) (first in the line of imported Python modules, hence the error mentioning tskit and not another Python module) in an R session which didn't have the correct Python environment activated.

Good luck!

---

I will leave this issue open as a reminder to link to the [detailed explanation](https://github.com/bodkan/slendr/issues/154#issuecomment-1997714497) somewhere in documentation. Probably after cleaning it up a bit and adding a new section between "Example" and "slendr specific example" which will prove the source of the error without bringing parallelized futures into the picture.

**Comment by bodkan** [comment]
Added a [brief description](https://www.slendr.net/articles/vignette-00-installation.html#modulenotfounderror-no-module-named-tskit-error) of the error with the link to the writeup above to the slendr website. I expanded it with a basic non-parallelized example of how the error happens. Together with the non-slendr pure R/Python example, this will hopefully become a useful resource to direct users to in case they run into the same problem.

Closing.
